### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by elyaishere**

### CODE REVIEW

File: src/concurrency/Executor.cpp [hunk 1]
```diff
@@ -1,5 +1,76 @@
...
+                threads.erase(t);
+            }
+    }
+
+    void Executor::Stop (bool await) {
+        std::unique_lock<std::mutex> lock(mutex);
+        state = State::kStopping;
+        if (await) {
+            while (!threads.empty()) {
+                empty_condition.wait(lock);
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:62]: Этот condition_variable _не про то_, что потоки завершились, он про то, что задача появилась

File: include/afina/concurrency/Executor.h [hunk 2]
```diff
@@ -51,11 +52,52 @@ class Executor {
...
 
         std::unique_lock<std::mutex> lock(this->mutex);
-        if (state != State::kRun) {
+        if (state != State::kRun || tasks.size() == _max_queue_size) {
             return false;
         }
-
-        // Enqueue new task
+        if (threads.size() == _working_threads && _working_threads < _high_watermark) {
+            threads.emplace_back([this, exec]() {
```
Comment (xphoenix) [inline include/afina/concurrency/Executor.h:59]: не надо дублировать такой код

File: src/concurrency/Executor.cpp [hunk 3]
```diff
@@ -1,5 +1,76 @@
...
+            while (true) {
+                std::function<void()> task;
+                {
+                    std::unique_lock<std::mutex> lock(mutex);
+                    if (tasks.empty()) {
+                        if (state == State::kRun) {
+                            auto wt = empty_condition.wait_for(lock, _idle_time);
+                            if (wt == std::cv_status::timeout && threads.size() > _low_watermark) {
+                                kill_thread();
+                                return;
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:24]: Лучше ```return``` по середине таких циклов не использовать

File: src/concurrency/Executor.cpp [hunk 4]
```diff
@@ -1,5 +1,76 @@
...
+        _working_threads = 0;
+        state = State::kRun;
+
+        for (size_t i = 0; i < _low_watermark; ++i) {
+            threads.emplace_back([this]() {
+            while (true) {
+                std::function<void()> task;
+                {
+                    std::unique_lock<std::mutex> lock(mutex);
+                    if (tasks.empty()) {
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:19]:
```
while (!tasks.empty() || state == kRun) {
if (tasks.empty() && state != kStopped) {
     status = wait(_lock);
     if (status == timeout) {
        continue;
     }
}
...
```

File: src/concurrency/Executor.cpp [hunk 5]
```diff
@@ -1,5 +1,96 @@
...
+
+void routine(Executor * pool) {
+    std::unique_lock<std::mutex> lock(pool->mutex);
+    while (!(pool->tasks.empty()) || (pool->state == Executor::State::kRun)) {
+                std::function<void()> task;
+                    
+                    // задач нет
+                    if (pool->tasks.empty()) {
+                        // ждем задачи
+                        auto wt = pool->empty_condition.wait_for(lock, pool->_idle_time);
```
Comment (Fen99) [inline src/concurrency/Executor.cpp:64]: Не учтен случай, если wait_for просыпается до таймаута.
```
while (!tasks.empty() || state == kRun) {
   wait_until_time = time::now() + idle_time;
   wait_result = ...;
   while (tasks.empty()) {
       empty_condition.wait_until(lock, wait_until_time);
       if (timeout) { wait_result = timeout; break; }
       else { if (tasks.empty()) { continue; } else { wait_result = new_task; } }  // Внезапное пробуждение
   }
   
   if (wait_result == timeout) { /* continue или завершение, если это лишний поток > low_watermark */}
   if (wait_result == new_task) { /* исполняем */
}

File: src/concurrency/Executor.cpp [hunk 6]
```diff
@@ -1,5 +1,96 @@
...
+    }
+
+    Executor::~Executor() {
+        Stop(true);
+    }
+
+void routine(Executor * pool) {
+    std::unique_lock<std::mutex> lock(pool->mutex);
+    while (!(pool->tasks.empty()) || (pool->state == Executor::State::kRun)) {
+                std::function<void()> task;
```
Comment (Fen99) [inline src/concurrency/Executor.cpp:59]: void() noexept

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: src/concurrency/Executor.cpp [hunk 7]
```diff
@@ -1,5 +1,110 @@
...
+    }
+
+    void Executor::kill_thread() {
+        auto id = std::this_thread::get_id();
+        threads.at(id).detach();
+        threads.erase(id);
+    }
+
+    void Executor::Stop (bool await) {
+        if (state == State::kStopped) {
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:27]: А как это без лока, да в общую память?

File: src/concurrency/Executor.cpp [hunk 8]
```diff
@@ -1,5 +1,110 @@
...
+                auto wait_result = new_task;
+                    // задач нет
+                    while (pool->tasks.empty()) {
+                        // ждем задачи
+                        auto wt = pool->empty_condition.wait_until(lock, wait_until_time);
+                        if (wt == std::cv_status::timeout){
+                            // время вышло
+                            wait_result = timeout;
+                            break;
+                        }else { // внезапное пробуждение
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:70]: весь этот else - абсолютно бесполезная хрень

File: src/concurrency/Executor.cpp [hunk 9]
```diff
@@ -1,5 +1,110 @@
...
+        if (await) {
+            std::unique_lock<std::mutex> lock(mutex);
+            while (!threads.empty()) {
+                stop_condition.wait(lock);
+            }
+        }
+        {
+            std::unique_lock<std::mutex> lock(mutex);
+            // чтобы раньше времени не остановиться
+            if (_working_threads == 0) {
```
Comment (xphoenix) [inline src/concurrency/Executor.cpp:45]: Нафига вот тут вот это, если оно выставляется в routine
